### PR TITLE
Boolean expression object

### LINKED ISSUE #5666: classical_function creates wrong circuits for some cases
### Information

- **Qiskit Terra version**: master
- **Python version**: 3.8
- **Operating system**: macOS Catalina

### What is the current behavior?
`classical_function` creates wrong circuits for some cases. I confirmed at least two cases. 

### Steps to reproduce the problem
Case 1) 
```
def expr1(x1: Int1, x2: Int1, x3: Int1) -> Int1:
    return x1

cf = classical_function(expr1)
print(cf.synth())

q_0: ──■──
       │  
q_1: ──┼──
       │  
q_2: ──┼──
     ┌─┴─┐
q_3: ┤ X ├
     └───┘
```

Case 2)
```
def expr2(x1: Int1, x2: Int1, x3: Int1) -> Int1:
    return x1 or x2 or x3

cf = classical_function(expr2)
print(cf.synth())

q_0: ──o────────────
       │            
q_1: ──o────o───────
       │    │       
q_2: ──■────■────■──
     ┌─┴─┐┌─┴─┐┌─┴─┐
q_3: ┤ X ├┤ X ├┤ X ├
     └───┘└───┘└───┘
```
### What is the expected behavior?
Case 1)
The control bit of the CNOT gate should be the most bottom bit. 

As written in the [documentation](https://qiskit.org/documentation/apidoc/classicalfunction.html), following Qiskit’s little-endian bit ordering convention, the left-most bit (x1) is the most significant bit and the right-most bit (x3) is the least significant bit. 
The following code is an example for the correct bit ordering.
```
def expr3(x1: Int1, x2: Int1, x3: Int1) -> Int1:
    return not x1 and x2 and x3

cf = classical_function(expr3)
print(cf.synth())

q_0: ──■──
       │  
q_1: ──■──
       │  
q_2: ──o──
     ┌─┴─┐
q_3: ┤ X ├
     └───┘
```

Case 2)
The circuit does not represent `x1 or x2 or x3`. For example, when (x1, x2, x3) = (001), $q_3$ of the circuit does not become 1.

### PR DESCRIPTION
**Pull request by 1ucian0**
fixes #5666

Based on https://github.com/Qiskit/qiskit-terra/pull/5656 (this extends and replace #5656), this PR:
 - implements `BooleanExpression` which:
   - leverages tweedledum python expression parser.
   - leverages tweedledum DIMACS parser.
 - creates `PhaseOracle` (similar to previous `LogicalExpressionOracle`) as a circuit wrapper of `BooleanExpression`

### CODE REVIEW

File: qiskit/algorithms/amplitude_amplifiers/amplification_problem.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 from qiskit.circuit import QuantumCircuit
 from qiskit.circuit.library import GroverOperator
 from qiskit.quantum_info import Statevector
+from qiskit.circuit.library.phase_oracle import PhaseOracle
```
Comment (Cryoris) [inline qiskit/algorithms/amplitude_amplifiers/amplification_problem.py:20]:
```suggestion
```

File: qiskit/circuit/classicalfunction/boolean_expression.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+from os.path import basename
+
+from qiskit.circuit import Gate
+from qiskit.exceptions import QiskitError
+from .classicalfunction import HAS_TWEEDLEDUM
+from .utils import tweedledum2qiskit
+
+
+class BooleanExpression(Gate):
+    r"""The Logical Expression Quantum Oracle.
```
Comment (Cryoris) [inline qiskit/circuit/classicalfunction/boolean_expression.py:24]: I think this docstring should be in the `PhaseOracle` class, since that's the replacement of the previous `LogicalExpressionOracle` where this docstring comes from 🙂

File: qiskit/circuit/classicalfunction/boolean_expression.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+                              '"pip install tweedledum".')
+        from tweedledum import BoolFunction
+
+        expr_obj = cls.__new__(cls)
+        expr_obj._tweedledum_bool_expression = BoolFunction.from_dimacs_file(filename)
+
+        super(BooleanExpression, expr_obj).__init__(
+            name=basename(filename),num_qubits=expr_obj._tweedledum_bool_expression.num_inputs() +
+                                               expr_obj._tweedledum_bool_expression.num_outputs(),
+            params=[])
```
Comment (Cryoris) [inline qiskit/circuit/classicalfunction/boolean_expression.py:172]: Alternatively, would it be more straightforward to do the following?
```python
expr = BoolFunction.from_dimacs_file(filename)
return cls(expr)
```

File: qiskit/circuit/library/__init__.py [hunk 4]
```diff
@@ -181,6 +181,7 @@
    QuantumVolume
    PhaseEstimation
    GroverOperator
+   LogicalExpressionOracle
```
Comment (Cryoris) [inline qiskit/circuit/library/__init__.py:184]:
```suggestion
   PhaseOracle
```

File: qiskit/circuit/library/__init__.py [hunk 5]
```diff
@@ -277,9 +278,6 @@
 
 from .standard_gates import *
 from .templates import *
-from ..barrier import Barrier
```
Comment (Cryoris) [inline qiskit/circuit/library/__init__.py:280]: Why are these removed?

**Review by Cryoris — DISMISSED** [review]
Approach LGTM 👍🏻

File: qiskit/circuit/library/__init__.py (hunk 5 above)
Comment (1ucian0) [reply qiskit/circuit/library/__init__.py]: good catch. Reverted in 8ede589fef1195006273b9f25a3c88473964d3b4

File: qiskit/circuit/classicalfunction/boolean_expression.py (hunk 3 above)
Comment (1ucian0) [reply qiskit/circuit/classicalfunction/boolean_expression.py]: this requires `from_dimacs_file` to create an expression, which needs to be parsed by constructor. I agree this `from_<somewhere>` is not nice.

File: qiskit/circuit/classicalfunction/boolean_expression.py (hunk 2 above)
Comment (1ucian0) [reply qiskit/circuit/classicalfunction/boolean_expression.py:22]: Yeap! Moved.

File: qiskit/circuit/library/phase_oracle.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    For convenience, this oracle, in addition to parsing arbitrary logical expressions,
+    also supports input strings in the `DIMACS CNF format
+    <http://www.satcompetition.org/2009/format-benchmarks2009.html>`__,
+    which is the standard format for specifying SATisfiability (SAT) problem instances in
+    `Conjunctive Normal Form (CNF) <https://en.wikipedia.org/wiki/Conjunctive_normal_form>`__,
+    which is a conjunction of one or more clauses, where a clause is a disjunction of one
+    or more literals. See :meth:`qiskit.circuit.library.phase_oracle.PhaseOracle.from_dimacs_file`.
+    """
+
+    def __init__(self, expression: str) -> None:  # pylint: disable=super-init-not-called
```
Comment (Cryoris) [inline qiskit/circuit/library/phase_oracle.py:38]: This structure with `_build_from_boolean_expression` seems a bit weirdly split to me. Couldn't we just have all of it in the `__init__`? Then we also wouldn't have to store `self.state_qubits` 🙂

**Review by Cryoris — COMMENTED** [review]
Looks good to me, I have just one comment around the class structure in `PhaseOracle` 👍🏻

File: qiskit/circuit/library/phase_oracle.py (hunk 6 above)
Comment (1ucian0) [reply qiskit/circuit/library/phase_oracle.py:38]: How the constructor call would look like?

Comment (Cryoris) [reply qiskit/circuit/library/phase_oracle.py:38]: hmm after trying to implement it, storing `boolean_expression` does make a lot of sense since you need it for the `evaluate` method. But we can still avoid storing `state_qubits` by rewriting `_build_from_boolean_expression` a bit:
```python
def _build_from_boolean_expression(self):
    qr_state = QuantumRegister(self.boolean_expression.num_qubits - 1, 'state')
    # rest as before
```
I'd prefer storing as little extra info as necessary, otherwise users might expect `state_qubits` available on other circuits.

Comment (1ucian0) [reply qiskit/circuit/library/phase_oracle.py:38]: oh! I thought `state_qubits` was part of the `PhaseOracle` API. Removed in 261ccc097

File: qiskit/circuit/classicalfunction/boolean_expression.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            expression (str): The logical expression string.
+            name (str): Optional. Instruction gate name. Otherwise part of
+                        the expression is going to be used.
+        Raises:
+            ImportError: If tweedledum is not installed. Tweedledum is required.
+        """
+        if not HAS_TWEEDLEDUM:
+            raise ImportError("To use the BooleanExpression compiler, tweedledum "
+                              "must be installed. To install tweedledum run "
+                              '"pip install tweedledum".')
```
Comment (1ucian0) [inline qiskit/circuit/classicalfunction/boolean_expression.py:36]:
```suggestion
                              '"pip install tweedledum==1.0.0b3".')
```

File: qiskit/circuit/classicalfunction/boolean_expression.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+            BooleanExpression: A gate for the input string
+
+        Raises:
+            ImportError: If tweedledum is not installed. Tweedledum is required.
+            FileNotFoundError: If filename is not found.
+        """
+        if not HAS_TWEEDLEDUM:
+            raise ImportError("To use the BooleanExpression compiler, tweedledum "
+                              "must be installed. To install tweedledum run "
+                              '"pip install tweedledum".')
```
Comment (1ucian0) [inline qiskit/circuit/classicalfunction/boolean_expression.py:110]:
```suggestion
                              '"pip install tweedledum==1.0.0b3".')
```

File: qiskit/circuit/classicalfunction/classical_function_visitor.py [hunk 9]
```diff
@@ -55,6 +52,12 @@ def visit_Module(self, node):
 
     def visit_FunctionDef(self, node):
         """The function definition should have type hints"""
+        if HAS_TWEEDLEDUM:
+            from tweedledum.classical import LogicNetwork  # pylint: disable=no-name-in-module
+        else:
+            raise ImportError("To use the classicalfunction compiler, tweedledum "
+                              "must be installed. To install tweedledum run "
+                              '"pip install tweedledum".')
```
Comment (1ucian0) [inline qiskit/circuit/classicalfunction/classical_function_visitor.py:60]:
```suggestion
                              '"pip install tweedledum==1.0.0b3".')
```

**Comment by 1ucian0** [comment]
This issue is blocked on recently found issues on tweedledum.

File: qiskit/circuit/library/phase_oracle.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    def __init__(self, expression: Union[str, ClassicalElement]) -> None:
+        if not isinstance(expression, ClassicalElement):
+            expression = BooleanExpression(expression)
+        super().__init__(expression.num_qubits - 1, name='Phase Oracle')
+        self.boolean_expression = expression
+
+        from tweedledum.passes import pkrm_synth  # pylint: disable=no-name-in-module
+        oracle = expression.synth(
+            synthesizer=lambda logic_network: pkrm_synth(logic_network,
+                                                         {"pkrm_synth": {"phase_esop": True}}))
```
Comment (1ucian0) [inline qiskit/circuit/library/phase_oracle.py:50]:
```suggestion
            synthesizer=lambda boolean_expression: pkrm_synth(boolean_expression._tweedledum_bool_expression.truth_table(output_bit=0),
                                                         {"pkrm_synth": {"phase_esop": True}}))
```

File: releasenotes/notes/bool_expression_phaseoracle-1802be3016c83fa8.yaml [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+      backend = BasicAer.get_backend('qasm_simulator')
+      
+      problem = AmplificationProblem(oracle, is_good_state=oracle.evaluate_bitstring)
+      grover = Grover(quantum_instance=backend)
+      result = grover.amplify(problem)
+      result.top_measurement
+
+  - |
+    The new method
+    :meth:`qiskit.algorithms.amplitude_amplifiers.amplification_problem.AmplificationProblem.from_dimacs_file`
```
Comment (Cryoris) [inline releasenotes/notes/bool_expression_phaseoracle-1802be3016c83fa8.yaml:82]: Hm I can't find this method, am I missing something or has this not yet been included? 🙂

File: releasenotes/notes/bool_expression_phaseoracle-1802be3016c83fa8.yaml [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+      oracle = PhaseOracle('x1 & x2 & (not x3)')
+      oracle.draw('mpl')
+    
+    These phase oracles can be used, for example, in 
+    :class:`qiskit.algorithms.amplitude_amplifiers.amplification_problem.AmplificationProblem`.
+
+    .. jupyter-execute::
+
+      from qiskit.algorithms.amplitude_amplifiers.amplification_problem import AmplificationProblem
+      from qiskit.algorithms.amplitude_amplifiers.grover import Grover
```
Comment (Cryoris) [inline releasenotes/notes/bool_expression_phaseoracle-1802be3016c83fa8.yaml:70]: The full paths are not needed, but if you wanted to use them for clarity that's also fine with me
```suggestion
      from qiskit.algorithms import AmplificationProblem, Grover
```

File: qiskit/circuit/library/phase_oracle.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+        :math:`(\lnot x_1 \lor \lnot x_2 \lor \lnot x_3)
+        \land (x_1 \lor \lnot x_2 \lor x_3)
+        \land (x_1 \lor x_2 \lor \lnot x_3)
+        \land (x_1 \lor \lnot x_2 \lor \lnot x_3)
+        \land (\lnot x_1 \lor x_2 \lor x_3)`.
+
+
+        Args:
+            filename (str): A file in DIMACS format.
```
Comment (Cryoris) [inline qiskit/circuit/library/phase_oracle.py:113]: The typehint is already in the signature
```suggestion
            filename: A file in DIMACS format.
```

File: qiskit/circuit/library/phase_oracle.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    which is the standard format for specifying SATisfiability (SAT) problem instances in
+    `Conjunctive Normal Form (CNF) <https://en.wikipedia.org/wiki/Conjunctive_normal_form>`__,
+    which is a conjunction of one or more clauses, where a clause is a disjunction of one
+    or more literals. See :meth:`qiskit.circuit.library.phase_oracle.PhaseOracle.from_dimacs_file`.
+
+    From 16 variables on, possible performance issues should be expected when using the
+    default synthesizer.
+    """
+
+    def __init__(self, expression: Union[str, ClassicalElement], synthesizer=None) -> None:
```
Comment (Cryoris) [inline qiskit/circuit/library/phase_oracle.py:44]: Can you add a typehint for the synthesizer? And a docstring for the arguments?

File: qiskit/circuit/classicalfunction/boolean_expression.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+    def _define(self):
+        """The definition of the boolean expression is its synthesis"""
+        self.definition = self.synth()
+
+    @classmethod
+    def from_dimacs_file(cls, filename: str):
+        """Create a BooleanExpression from the string in the DIMACS format.
+        Args:
+            filename (str): A file in DIMACS format.
```
Comment (Cryoris) [inline qiskit/circuit/classicalfunction/boolean_expression.py:97]:
```suggestion
            filename: A file in DIMACS format.
```

File: qiskit/circuit/classicalfunction/boolean_expression.py [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+            truth_table = self._tweedledum_bool_expression.truth_table(output_bit=0)
+            return tweedledum2qiskit(pkrm_synth(truth_table), name=self.name, qregs=qregs)
+        return synthesizer(self)
+
+    def _define(self):
+        """The definition of the boolean expression is its synthesis"""
+        self.definition = self.synth()
+
+    @classmethod
+    def from_dimacs_file(cls, filename: str):
```
Comment (Cryoris) [inline qiskit/circuit/classicalfunction/boolean_expression.py:94]: We could also use the same structure as in `PhaseOracle` to clean up the alternative constructor. But I'm also good with this one 🙂

File: qiskit/circuit/classicalfunction/boolean_expression.py [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+        Returns:
+            bool: result of the evaluation.
+        """
+        from tweedledum import BitVec
+        bits = []
+        for bit in bitstring:
+            bits.append(BitVec(1, bit))
+        return bool(self._tweedledum_bool_expression.simulate(*bits))
+
+    def synth(self, registerless=True, synthesizer=None):
```
Comment (Cryoris) [inline qiskit/circuit/classicalfunction/boolean_expression.py:65]: Could you move the typehints to the signature?

File: qiskit/circuit/classicalfunction/boolean_expression.py [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+
+from qiskit.exceptions import MissingOptionalLibraryError
+from .classical_element import ClassicalElement
+from .utils import HAS_TWEEDLEDUM
+
+
+class BooleanExpression(ClassicalElement):
+    """The Boolean Expression gate."""
+
+    def __init__(self, expression: str, name: str = None) -> None:
```
Comment (Cryoris) [inline qiskit/circuit/classicalfunction/boolean_expression.py:25]: `name` should have the type hint `Optional[str] = None` to show that `None` is also a valid value

File: qiskit/circuit/classicalfunction/boolean_expression.py [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+            MissingOptionalLibraryError: If tweedledum is not installed. Tweedledum is required.
+        """
+        if not HAS_TWEEDLEDUM:
+            raise MissingOptionalLibraryError(
+                libname='tweedledum',
+                name='BooleanExpression compiler',
+                pip_install='pip install tweedledum')
+        from tweedledum import BoolFunction
+        self._tweedledum_bool_expression = BoolFunction.from_expression(expression)
+        self.qregs = None  # TODO: Probably from self._tweedledum_bool_expression._signature
```
Comment (Cryoris) [inline qiskit/circuit/classicalfunction/boolean_expression.py:41]: Is this variable needed?

**Review by Cryoris — COMMENTED** [review]
Mostly LGTM, just a few comments!

File: releasenotes/notes/bool_expression_phaseoracle-1802be3016c83fa8.yaml (hunk 11 above)
Comment (1ucian0) [reply releasenotes/notes/bool_expression_phaseoracle-1802be3016c83fa8.yaml:82]: Fixed in 8988e1c36

**Comment by 1ucian0** [comment]
tweedledum 1.0.0 released https://pypi.org/project/tweedledum/1.0.0/
removing `on hold`

File: qiskit/circuit/library/phase_oracle.py (hunk 14 above)
Comment (1ucian0) [reply qiskit/circuit/library/phase_oracle.py:44]: Fixed in 9bd5e3493

File: qiskit/circuit/classicalfunction/boolean_expression.py (hunk 17 above)
Comment (1ucian0) [reply qiskit/circuit/classicalfunction/boolean_expression.py]: done in c025c487d

File: qiskit/circuit/classicalfunction/boolean_expression.py (hunk 16 above)
Comment (Cryoris) [reply qiskit/circuit/classicalfunction/boolean_expression.py:100]: ok 👍🏻

File: qiskit/circuit/classicalfunction/boolean_expression.py (hunk 18 above)
Comment (1ucian0) [reply qiskit/circuit/classicalfunction/boolean_expression.py:29]: done in 15f48136d

File: qiskit/circuit/classicalfunction/boolean_expression.py (hunk 19 above)
Comment (1ucian0) [reply qiskit/circuit/classicalfunction/boolean_expression.py]: removed in e36abbd30

**Review by Cryoris — DISMISSED** [review]

File: qiskit/circuit/classicalfunction/boolean_expression.py [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+        Returns:
+            bool: result of the evaluation.
+        """
+        from tweedledum import BitVec
+        bits = []
+        for bit in bitstring:
+            bits.append(BitVec(1, bit))
+        return bool(self._tweedledum_bool_expression.simulate(*bits))
+
+    def synth(self, registerless: bool =True,
```
Comment (Cryoris) [inline qiskit/circuit/classicalfunction/boolean_expression.py:68]:
```suggestion
    def synth(self, registerless: bool = True,
```

**Review by Cryoris — DISMISSED** [review]

**Review by Cryoris — DISMISSED** [review]
